Request-a-copy improvements: Support access via secure links

kshepherd opened the pull request:
## References
* Requires [dspace-angular#3984](https://github.com/DSpace/dspace-angular/pull/3984)
* Related to REST contract: https://github.com/DSpace/RestContract/pull/307

## Description
Request a copy is an existing feature in DSpace, whereby users can request access to files they are not authorized to download, and the item author or a delegated helpdesk can approve or deny these requests.

Up until 8.x, all approved requests resulted in an email sent to the requester as attachments.

This is not a suitable solution for large files and doesn’t allow as much control over future authorization to those resources.

In this improvement, a secret key is generated to provide limited access to the requested bitstream(s) with a URL, rather than sending files as attachments. This link can be configured to allow access to just selected bitstreams, or all bitstreams of an item, and it can be configured with an “access period”, after which the secret URL will no longer work.

Example "access by secure link" item page:
![screenshot_2025-02-13_15:31:09_selection](https://github.com/user-attachments/assets/85ebc531-155d-4bb1-ab82-c63216d855f1)

## Notes for Reviewers
To test, you will also need the dspace-angular pull request at: https://github.com/DSpace/dspace-angular/pull/3984

This is halfway between an improvement and a new feature. For those doing practical testing, I recommend reading the attached PDF: [request-a-copy-secure-links.pdf](https://github.com/user-attachments/files/18785739/request-a-copy-secure-links.pdf)

**There is a flyway sql migration** - two columns are added, so make sure you are testing on an instance that doesn't need to quickly roll back to an earlier version with the existing requestitem table.

**There is a new captcha added** - A new GDPR complicant, proof-of-work captcha is added to the "request a copy" form. I would like to propose we make this the default for Register User as well, and am happy to open a new PR accordingly. It is under the MIT license: https://altcha.org/captcha/

## Instructions for Reviewers
1. Read the supplied PDF: 
[request-a-copy-secure-links.pdf](https://github.com/user-attachments/files/18785739/request-a-copy-secure-links.pdf)

2. Check out or merge this PR into your local environment, edit configuration to suit the way you wish to test, and build + deploy
    a. If you are a non-developer tester, you may be able to follow the docker testing instructions instead
3. Make some items with a mix of restricted and public files, of various sizes (depending on the threshold you set, if any)
4. As anonymous and/or authenticated (but not authorized) users, make requests for the bitstreams or 'all files of items' in question. Take note of how the request form operates, ensuring that the captcha works well (it does not require human interaction).
5. As the owner or helpdesk delegate, check for the item request (if you are using a test environment and have mail server disabled in your config, you can see the email and copy the link from your log files). Follow the link to the approve/deny form
6. Try approvals and denials for requests as an owner or helpdesk delegate, with different access periods.
7. Check emails (or logs if disabled) for both kinds of emails, depending on your configuration. You may have some smaller files sent as attachments, larger ones sent as a secure link instead, you may have everything configured to send as a link, and so on. 
8. Test out access to the bitstreams / item page using the secure link provided. Note the behaviour of downloading a bitstream explicitly allowed by this link, compared to a bitstream the user already had access to, compared to a bitstream that is still unauthorized (and which should lead to a new request copy form)
9. Try accessing files with the link outside of the access period
10. Try images and videos with the media viewer enabled

List of changes in this PR:
1. New columns added to request item model (access period, access token)
2. RequestItemService and RestRepository changes to support secure access by token
3. BitstreamResourceAccessByToken to support access to a bitstream using a secret query param string
4. New Altcha captcha (https://altcha.org/captcha/)
5. New email template to support secure token

## Attribution
Developed by The Library Code with the support of Technische Universität Berlin

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [ ] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [ ] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.

kshepherd committed:
Request-a-copy improvement: additional configuration

kshepherd committed:
Request-a-copy improvement: Package info and POM changes

kshepherd committed:
Request-a-copy improvement: New configuration

kshepherd committed:
Request-a-copy improvement: Email template

kshepherd committed:
Request-a-copy improvement: SQL migration scripts

kshepherd committed:
Request-a-copy improvement: Altcha Recaptcha service

kshepherd committed:
Request-a-copy improvement: Core model, DAO, service changes

kshepherd committed:
Request-a-copy improvement: Email notifier changes

kshepherd committed:
Request-a-copy improvement: REST server changes

kshepherd committed:
Request-a-copy improvement: Altcha token header allowed

kshepherd committed:
Request-a-copy improvement: Unit and integration tests

kshepherd commented:
Regarding the failing integration tests - these are because I replaced the Google Recaptcha (for now) in the test spring services XML, with Altcha.

Happy to have a discussion about whether we run with one or the other, or even if I just split my work off so it is not implementing the same interface and so doesn't interfere

kshepherd committed:
Request-a-copy: Merge recaptcha headers to generic 'x-captcha-payload'

This helps with multiple providers

kshepherd committed:
Request-a-copy: Further captcha config changes

MW3000 commented:
I tested the main functionality and it works great and as advertised ✅. I will test the different configuration possibilities in the next days.

kshepherd committed:
Request-a-copy: Support dual captcha spring services

MW3000 commented:
I did thoroughly test the different configuration options:

# Test basic configuration
- Request all files (of item) in restricted access, grant access => Works ✅
- Request one of many files, grant access => Works ✅
- Deny access request => Works ✅

# Test the captcha
=> Works ✅
(don't forget to set `altcha.hmac.key` when testing)

# Test using a size limit
Request single file below limit
=> Works ✅
Request single file above limit
=> Works ✅
If 'all files' was indicated in the request, this threshold will be activated if any of the files meet the configured size minimum.
=> Works ✅

# Test if access is closed again, after end of access period
=> Works insofar as access is no longer possible ✅
💡 But, after the end of the access period, I just get presented with a login page. It might be a better UX to show the custom download page and add a hint, that the access period expired?

# Test restricted images and videos with media viewer enabled
=> Image Viewer works on download page ✅
💡 Should be ported to work on regular item page too! (unrelated to this issue! but we could perhaps use the solution from this pull request?)

=> Video player works on download page ✅
💡 Should be ported to work on regular item page too! (unrelated to this issue! but we could perhaps use the solution from this pull request?)

tdonohue commented on dspace-api/src/test/java/org/dspace/app/requestitem/RequestItemTest.java:
```diff
@@ -0,0 +1,364 @@
+
+        request.setAccept_request(true);
+        request.setDecision_date(new Date(System.currentTimeMillis()));
+        request.setAccess_period(7); // 7 day access
```
This is a 7 **second** access period, isn't it?  The comment is just wrong here, unless you meant to multiply `7 * 86400`

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/AltchaCaptchaRestController.java:
```diff
@@ -0,0 +1,181 @@
+ *
+ * @author Kim Shepherd
+ */
+@RequestMapping(value = "/api/captcha")
```
We need to document these new endpoints in our REST Contract https://github.com/DSpace/RestContract/blob/main/README.md

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/RequestCopyFeature.java:
```diff
@@ -65,6 +65,7 @@ public class RequestCopyFeature implements AuthorizationFeature {
     @Override
     public boolean isAuthorized(Context context, BaseObjectRest object) throws SQLException {
         String requestType = configurationService.getProperty("request.item.type");
+        log.warn("isAuthorized for REQ A COPY IS requestType=" + requestType);
```
This looks like something you were using for debugging and should be removed.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/BitstreamResourceAccessByToken.java:
```diff
@@ -0,0 +1,203 @@
+     * @throws AuthorizeException
+     */
+    private Context initializeContext() throws SQLException, AuthorizeException {
+        return new Context();
```
We might consider making this a READ-ONLY context.  `new Context(Mode.READ_ONLY)`.

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/RequestItemRepositoryIT.java:
```diff
@@ -136,6 +144,8 @@ public void init()
     @Test
     public void testFindAll()
             throws Exception {
+        System.out.println("findAll");
```
I believe our ITs will already print out the method name. I don't think these `System.out.println()` are necessary, as you should already see logs that say something like  "Running testFindAll()"  (I'm forgetting the exact log message though)

tdonohue commented on dspace/config/emails/request_item.granted_token:
```diff
@@ -0,0 +1,36 @@
+    ${params[2]}
+    ${params[1]}
+
+#if( $params[7] )
```
Parameters 6 & 7 should be described above in the comment.

tdonohue commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,43 @@
+#request.item.reject.email = true
+# Bundles to use when granting access and inspecting file size, if "all files" is indicated
+# Default: ORIGINAL
+#request.item.grant.bundles = ORIGINAL
```
Is this a comma separated list?  We should document how to add multiple.

tdonohue commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,43 @@
+# By default a granted request will send the file as an attachment.
+# If a secure web link to the bitstream(s) should be sent in some cases, enable the following property
+# Default: false
+#request.item.grant.link = true
```
I think we should better describe this setting instead of saying "in some cases".  We should say that the link is sent if the files are above a specific size (specified in "filesize") below.

tdonohue commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,43 @@
+#request.item.grant.link.period = 604800
+# Require a captcha for item request creation
+# Default: false
+#request.item.create.captcha=true
```
Just a general comment.  It's very difficult to read this file as there are no empty lines between each setting.  We may want to separate each setting a bit better to make it more readable.

tdonohue commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,43 @@
+# Valid access periods, in seconds, to allow the approver to select when using links.
+# These are presented to users with friendly labels via i18n keys in the angular frontend.
+# A 0 (or null) access period will be interpreted as 'forever'.
+# The first access period in the list will be the default option in the grant form dropdown.
```
Are there any defaults?  If so, we should document the default values.  If there are no defaults, then we may want to uncomment the below list.  Sites may expect this to "just work" if they set `request.item.grant.link = true`.

kshepherd replied:
I've updated the defaults along with the storage period. they are uncommented since the feature is disabled by default so there is no harm keeping these open
```
request.item.grant.link.period = FOREVER
request.item.grant.link.period = +1DAY
request.item.grant.link.period = +1WEEK
request.item.grant.link.period = +1MONTH
request.item.grant.link.period = +3MONTHS
```

tdonohue commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,43 @@
+#request.item.grant.link.period = 86400
+#request.item.grant.link.period = 604800
+# Require a captcha for item request creation
+# Default: false
```
Should this default to "true" since it's using a semi-"built-in" captcha approach?

Does the captcha also work if you just use the traditional request-a-copy?  Or does it require enabling `request.item.grant.link = true`?

kshepherd replied:
It will protect the form regardless of the sharing method, yes. I'll make sure the defaults are all ok, there might be something I can do to further separate it from google recaptcha, too (since i abandoned my "combine captcha" work for lack of time)

tdonohue reviewed (changes requested):
@kshepherd : Thanks for this work!  I haven't had a chance to test this yet, but I have done a code review.  Overall, the code looks great, but I do have some minor suggestions/comments inline (many of these are small requests, or notes to consider, but *not* requirements).

I do see that the automated tests are failing here because of the new Altcha Captcha.  How are you planning to fix that?  Are you just waiting on feedback before fixing the broken tests to also work for Altcha Captcha? (It looks like the failures are only occurring in `RegistrationRestRepositoryIT`).

ybnd commented on dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemEmailNotifier.java:
```diff
@@ -188,34 +197,48 @@ public void sendResponse(Context context, RequestItem ri, String subject,
+                    // {6} secure access link
+                    email.addArgument(configurationService.getProperty("dspace.ui.url")
+                            + "/items/" + ri.getItem().getID()
+                            + "/access-by-token?accessToken=" + ri.getAccess_token());
```
This doesn't match the latest path from https://github.com/DSpace/dspace-angular/pull/3984 anymore, I think you can remove `/access-by-token`

kshepherd replied:
@ybnd I've pushed a change that accepts either the +7DAYS string delta format, or a valid date string, and uses that (and if delta, the decision date) to store an accessExpiry field as timestamp in the database.

(note this means the sql migration script has changed)

ybnd reviewed (changes requested):
Haven't gone in-depth yet, but here are some small comments

kshepherd committed:
Request-a-copy: Access expiry sent as delta or date, stored as date

kshepherd commented on dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V9.0_2025.02.13__add_access_colums_to_requestitem_table.sql:
```diff
@@ -0,0 +1,13 @@
+ALTER TABLE requestitem ADD COLUMN IF NOT EXISTS access_token VARCHAR(48);
+-- Add new access_period column to hold a time delta in seconds (from decision_date timestamp) to calculate validity
+-- and expiry, with NULL interpreted as 'forever' (if accept_request is true). int4 allows for 68 year period max.
+ALTER TABLE requestitem ADD COLUMN IF NOT EXISTS access_period INT4;
```
@ybnd @tdonohue I've pushed a change that accepts either the +7DAYS string delta format, or a valid date string, and uses that (and if delta, the decision date) to store an accessExpiry field as timestamp in the database.

(note this means the sql migration script has changed)

kshepherd commented on dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemServiceImpl.java:
```diff
@@ -128,4 +205,148 @@ public boolean isRestricted(Context context, DSpaceObject o)
+                            && authorizeService.isAdmin(context, requestItem.getItem())) {
+                        // Return original object, this person technically had full access to the request item data via
+                        // the original approval link
+                        log.debug("User is authorized to receive all request item data: {}", currentUser.getEmail());
```
thanks, removed - old check before i realised i can ALWAYS sanitize as the admin/approver does not need access to this information after approval or denial

kshepherd commented on dspace-api/src/main/java/org/dspace/app/requestitem/RequestItem.java:
```diff
@@ -192,4 +199,68 @@ public Date getRequest_date() {
+     * @return A unique token to be used by the requester when granted access to the resource, which
+     * can be emailed upon approval
+     */
+    public String getAccess_token() {
```
I've updated the comments so hopefully it is clear that the first getToken() is for the approver/helpdesk to access the grant/deny form, and the access token is for the requester, after it is approved

kshepherd commented on dspace-api/src/main/java/org/dspace/app/requestitem/RequestItem.java:
```diff
@@ -192,4 +199,68 @@ public Date getRequest_date() {
+            calendar.setTime(decision_date);
+            calendar.add(Calendar.SECOND, access_period);
+            // Return boolean result of "access period end date is AFTER now"
+            return calendar.getTime().after(new Date());
```
this code is removed after the refactor to date/time storage for access expiry

kshepherd commented:
@ybnd @tdonohue @paulo-graca thanks for the feedback so far - I pushed an initial refactor to date/time storage, i expect some codestyle and unit/IT failures, but hopefully it shows you the working version.

the frontend PR is similarly updated and i will aim to get tests and style fixed by tomorrow morning.

kshepherd committed:
Request-a-copy: Config doc fixes

paulo-graca reviewed (commented):
Thank you @kshepherd for adding this to DSpace! It's a very wanted feature from our community.

I wanted to test it (with yesterday's code), following all the required steps. But currently I'm unable to do it. 
I'm using a small repository with near 500 items, with some restricted items. When I try to access a restricted any item with the anonymous user, the login page comes up. I wasn't expecting it. I was hopping to see the request item form. I'm missing something?

I'm also using the related angular PR.

Other comments: I agree with earlier comments, a date is preferable way to store the token's expire date and adding more readability to the conf file. And when the PR merge occurs, also, a wiki page should also be provided because, by default, ReCAPTCHA isn't needed anymore. This message should be clear for all. I say this because ReCAPTCHA is a concern we have when configuring a new repository.

kshepherd commented:
@paulo-graca thanks for testing - could you share your config (requestitem, altcha) so I can take a look?
i've just come down with (another) illness but hoping to get more progress on this ASAP as we near deadlines and all the refactors / changes needed, mean i need to rewrite all the tests

paulo-graca commented:
I wasn't changing any of the default settings.

paulo-graca commented:
these are my cookies
![image](https://github.com/user-attachments/assets/2c3b6d97-88dd-4b1f-864b-4b2a429ae859)

paulo-graca commented:
Following all the process I'm getting a 404 error when trying to download the requested bitstream:

Request:
GET https://[HOST]/server/api/tools/itemrequests/search/byAccessToken?accessToken=3D126931c8fb3f7fc687034f81bc6e8082
returns 404 error

with:
```
{
    "timestamp": "2025-03-21T15:56:22.506+00:00",
    "status": 404,
    "error": "Not Found",
    "message": "An exception has occurred",
    "path": "/server/api/tools/itemrequests/search/byAccessToken"
}
```

And with log message:
```
2025-03-21 15:56:22,505 WARN  7cec8517-1ad2-440e-ba18-0557e4ad5921 3043cd10-d7f6-43e0-90bb-4742b0ecc2ec org.dspace.app.rest.exception.DSpaceApiExceptionControllerAdvice @ An exception has occurred (status:404 exception: No such request item for accessToken=3D126931c8fb3f7fc687034f81bc6e8082 at: org.dspace.app.rest.repository.RequestItemRepository.findByAccessToken(RequestItemRepository.java:362))
```

I've just noticed that at the DB table, we have this token: 3D126931c8fb3f7fc687034f81bc6e8082
but in the token column instead of the access_token column. When I manually changed request URL with the access token that it's in the same row:

https://[HOST]/server/api/tools/itemrequests/search/byAccessToken?accessToken=3D126931c8fb3f7fc687034f81bc6e8082

Now, I access the item:
![image](https://github.com/user-attachments/assets/530f5cb9-e4a0-49d0-8f75-efd707da0544)

(probably this is an issue with the email template)

paulo-graca commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,59 @@
+# If a secure web link to the bitstream(s) should be sent when files are over a certain file
+# size threshold, enable the following property
+# Default: false
+#request.item.grant.link = true
```
My opinion is that this should be the default behavior. A link should be send by default.

kshepherd replied:
very happy to change this based on combined reviewer feedback, i just tend to a more conservative approach of not enabling new features by default

tdonohue replied:
I agree with @paulo-graca that we may want to default this to `true`.  The reason is that current behavior is *not* ideal because large files cannot always be attached to an email.  So, a default of `true` provides improved behavior as that solves that issue.

paulo-graca commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/RequestItemRepository.java:
```diff
@@ -195,7 +238,7 @@ public RequestItemRest createAndReturn(Context ctx)
         String responseLink;
         try {
-            responseLink = getLinkTokenEmail(ri.getToken());
+            responseLink = requestItemService.getLinkTokenEmail(ri.getToken());
```
I'm not sure if this should be the accessToken. Please check my comments from my test results.

kshepherd replied:
No. The approver gets the token that has always existed in this code (hence the old name "token" without additional context). The access token is for the email that is emailed to the requester after approval. This code is handling the email sent to an approver.

paulo-graca reviewed (changes requested):
@kshepherd I hope you get better soon!
This is a great work and I hope it could be merged soon. It would be a very welcomed feature. 
I found two tiny issues that should be addressed. The first is that there should be default settings that enable request a copy to work as expected. 
The other one is related with the Token that is send in the email. It should be the access token, but the token is sent instead. 
Now I understand @tdonohue concerns about the naming of the access token and also having a token in the same object. This might create confusion to other contributors and be a potential source of future misleading errors. Good commenting is vital to help others.
Addressing this two issues I'm ok with the merge of this PR.

kshepherd commented on dspace/config/emails/request_item.granted_token:
```diff
@@ -0,0 +1,36 @@
+
+Your request for a copy of the file(s) from the below document has been approved.  The following link will grant access to the files you requested.
+
+Secure link: ${params[6]}
```
THere seems to be trouble with teh expiry date setting on PUT although it was testing well earlier so I'm working on that now..

kshepherd committed:
Request-a-copy: Use postgres max timestamp, not Instant.MAX

kshepherd committed:
Request-a-copy: Set better defaults

kshepherd commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,59 @@
+# all - Anonymous users can request an item
+# logged - Login is mandatory to request an item
+# empty/commented out - request-copy not allowed
+#request.item.type = all
```
yes thanks, good catch, i updated this, helpdesk.override and reject.email so they are uncommented as per their values from dspace.cfg in main branch

kshepherd commented on dspace/config/dspace.cfg:
```diff
@@ -1676,6 +1662,7 @@ include = ${module_dir}/openaire-client.cfg
 include = ${module_dir}/orcid.cfg
 include = ${module_dir}/qaevents.cfg
 include = ${module_dir}/rdf.cfg
+include = ${module_dir}/requestitem.cfg
```
yes, i should not have commented out `request.item.type = all`, fixed as per below messages

kshepherd committed:
Request-a-copy: Fixes to h2 migration and test cfg

kshepherd committed:
Request-a-copy: Shift min/max timestamp getters to Utils

kshepherd committed:
Request-a-copy: Update / tidy BitstreamResource per coverpage changes

kshepherd committed:
Request-a-copy: Config and lint fixes

kshepherd commented:
@tdonohue @ybnd @paulo-graca thanks all for your feedback and reviews so far. This is ready for a another review as I believe I have addressed all feedback. I have significantly rewritten and simplified a lot of areas:
* Instead of storing access_period as a 4-bit integer representing seconds from decision date, a timestamp `access_expiry` column is used instead. I did not want to use NULL as "forever" as that seemed unsafe / prone to error, so instead the maximum date for that field is used (new `getMaxTimestamp` and `getMinTimstamp` static methods are added to `Util` to help easily work with these postgres min and max timestamp values)
* Altcha and Google captcha services still implement the same interface but a new `CaptchaServiceFactory` uses qualifier annotations and config to return the correct service when `getCaptchaService()` is called (this avoids spring conflicts in exposing both via core services)
* Other feedback addressed (inline doc, code style, code cleanup, etc)
* Other changes are in the frontend - in short, the simplfiication of obtaining request item data and implementing the changes in the core components rather than duplicating a lot of similar components

Note: all unit and integration tests are passing as of commit [`ce71fcf`](https://github.com/DSpace/DSpace/commit/ce71fcfd79ae6033f9ac586bd969d7e65547e2fb) but the checks are incorrectly "failed" due to a spurious CodeQL failure, so please treat checks as 'passed' for this purpose.

tdonohue commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,59 @@
+
+# Date format to use for the access expiry date in the grant email
+request.item.grant.link.dateformat = dd/MM/yyyy HH:mm:ss
+#request.item.grant.link.dateformat = MM/dd/yy
```
I'm not sure this example will work, as it doesn't have a *time*.  It may need to be `MM/dd/yyyy HH:mm:ss` (or similar)

tdonohue reviewed (changes requested):
@kshepherd : Gave this another review & test today.  Overall, the code looks great.  I do have a few *minor* recommendations inline below.  I don't think any of these are significant though, and let me know if you see a problem with any.

As for testing, I've verified the basic feature is working.  However, I **cannot download a restricted file with the access token**.  Here's what I'm doing/seeing:

1. First, I install both this PR and the frontend PR.   I'm running the frontend in production mode (`npm run build:prod; npm run serve:ssr`)
2. In my local.cfg, I've set these:
```
request.item.grant.link = true
request.item.grant.link.dateformat = yyyy-MM-dd HH:mm:ss
request.item.grant.link.filesize = 10
```
3. Find an Item with a restricted bitstream. In my case, I found one with a file that was 12MB in size. (That's why I set the filesize to 10MB)
4. Click on file while not logged in.  Submit a request for the file.
5. The email is sent, and I grabbed the link in the email to approve the request.
6. After approving the request, I see that the email is sent with the secure link
7. In an Incognito window, I opened up the secure link which looks like this: `http://localhost:4000/items/9a171726-3060-4859-9d0d-8c147cb148c5?accessToken=b93bbfc488d00aaaae1ebc96f4433b16`
8. The Item page loads with the note at the top about having secure link access.  The restricted file now shows an unlocked icon.
9. Click on the restricted file link (which has the same accessToken: `http://localhost:4000/bitstreams/c86d3265-782e-49a5-8226-014e75e44982/download?accessToken=b93bbfc488d00aaaae1ebc96f4433b16`)
10. I see an immediate 500 error page.  The Java Stacktrace looks like this:
```
2025-03-25 21:29:24,119 ERROR unknown 08c3465f-7e65-4b4a-8081-e022849d04e4 org.dspace.app.rest.exception.DSpaceApiExceptionControllerAdvice @ An exception has occurred (status:500)
java.lang.RuntimeException: org.dspace.authorize.AuthorizeException: Authorization denied for action READ on BITSTREAM:c86d3265-782e-49a5-8226-014e75e44982 by user null
  at org.dspace.app.rest.utils.BitstreamResource.fetchDocument(BitstreamResource.java:135) ~[dspace-server-webapp-9.0-SNAPSHOT.jar!/:9.0-SNAPSHOT]
  at org.dspace.app.rest.utils.BitstreamResource.getChecksum(BitstreamResource.java:111) ~[dspace-server-webapp-9.0-SNAPSHOT.jar!/:9.0-SNAPSHOT]
   at org.dspace.app.rest.BitstreamRestController.retrieve(BitstreamRestController.java:206) ~[dspace-server-webapp-9.0-SNAPSHOT.jar!/:9.0-SNAPSHOT]
   at jdk.internal.reflect.GeneratedMethodAccessor216.invoke(Unknown Source) ~[?:?]
   ...
Caused by: org.dspace.authorize.AuthorizeException: Authorization denied for action READ on BITSTREAM:c86d3265-782e-49a5-8226-014e75e44982 by user null
   at org.dspace.authorize.AuthorizeServiceImpl.authorizeAction(AuthorizeServiceImpl.java:172) ~[dspace-api-9.0-SNAPSHOT.jar!/:9.0-SNAPSHOT]
   at org.dspace.authorize.AuthorizeServiceImpl.authorizeAction(AuthorizeServiceImpl.java:120) ~[dspace-api-9.0-SNAPSHOT.jar!/:9.0-SNAPSHOT]
   at org.dspace.authorize.AuthorizeServiceImpl.authorizeAction(AuthorizeServiceImpl.java:114) ~[dspace-api-9.0-SNAPSHOT.jar!/:9.0-SNAPSHOT]
   at org.dspace.content.BitstreamServiceImpl.retrieve(BitstreamServiceImpl.java:303) ~[dspace-api-9.0-SNAPSHOT.jar!/:9.0-SNAPSHOT]
   at org.dspace.app.rest.utils.BitstreamResource.fetchDocument(BitstreamResource.java:132) ~[dspace-server-webapp-9.0-SNAPSHOT.jar!/:9.0-SNAPSHOT]
  ... 195 more
```

Is this PR working for you?  Can you see anything that I might have done wrong?  Beyond that final download request, everything else looks good to me.

kshepherd commented:
@tdonohue yes you're right about the download issue - introduced when i was pushing the last few changes where i didn't realise that the changed BitstreamResource (recently changed in main) I extend was trying to retrieve the bitstream the "normal" way for checksum / content-length getters, that will be fixed as I push the toher changes to your feedback.

Captcha endpoints for the REST contract can be found at https://github.com/DSpace/RestContract/pull/307

kshepherd commented on dspace-api/src/test/data/dspaceFolder/config/local.cfg:
```diff
@@ -188,6 +188,13 @@ ldn.ip-range.enabled = true
+request.item.create.captcha = false
+request.item.type = all
+altcha.hmac.key = onetwothreesecret
+captcha.provider = google
```
Yes, in particular now that we are enabling more of the feature by default, all I need is the HMAC key to be manually set in the classes that need to test the ALTCHA-protected forms, I have removed all my test config overrides.

kshepherd committed:
Request-a-copy: Review feedback changes

Enabled feature by default
Test configs removed from local.cfg
Inline docs updated
Tests updated
BitstreamResource issues fixed

kshepherd commented:
@tdonohue Thanks for the prompt review! I have addressed your points, and the recent issue with BitstreamResource download authorization. Furthermore, that side-effect *should* have been tested for since it is fundamental to the feature! So I realised that actual bitstream download test coverage was not adequate, and added more tests to `BistreamControllerIT`

* Fixed BitstreamResource download issue and wrote new IT tests
* Removed local.cfg test overrides, set all required properties in Test classes themselves
* Activated the feature by default (but not the captcha, as it requires a secret to be configured as noted)
* Tidied code and comments as per feedback

As noted I think the existing RestContract PR covers the endpoints ok: https://github.com/DSpace/RestContract/pull/307

tdonohue reviewed (commented):
@kshepherd: This is looking good overall, and I'm almost +1.  In fact, this review *started* as a +1 until I stumbled on one minor bug.

**If you select the `+1WEEK` option when granting a request copy, then the resulting `accessToken` is INVALID.** The problem appears to be that the `access_expiry` date ends up being `null` in the database.  I'm not exactly sure why though.

Everything else about this PR looks good.  I've tested *every other* default period (Forever, 1 day, 1 month, 3 months) and they all work.  It's only the "1 week" period that fails.

I've verified the download of the bitstream using the `accessToken` is now working.   I also tested setting the expiration date of an existing access token to a date/time in the past, and verified that the `accessToken` no longer works once expired.  

Finally, I've tested turning on Altcha and verified it is working well also.

So, I'm **very close** to a +1, except for that `+1WEEK` bug.  If we cannot figure out a quick fix though, then another option is to remove it and document that "WEEK" is not yet supported (but I'm hoping maybe it's some small bug).

tdonohue commented:
@ybnd and @paulo-graca :  If one of you could re-review, this feature is now working mostly for me (except for the `+1WEEK` option as noted above).  I believe all your prior feedback has been addressed/fixed, and I'd like to get this merged soon.

(Please note that the CodeQL error on this PR should be disregarded, as it's a false positive.  It's noting a flaw in unrelated code which exists already on `main`, so it's unrelated to this PR's new code.)

kshepherd committed:
Request-a-copy: +1WEEK -> +7DAYS delta fix

tdonohue commented:
> @tdonohue you are totally right, our DateMathParser actually doesn't support that unit word! DAY, MONTH and YEAR are the units it supports. 

@kshepherd : Aha, that makes sense then.  In that case, don't forget to fix/update the configuration instructions in `requestitem.cfg` for `request.item.grant.link.period` to make it clear that only DAYS, MONTHS and YEARS are supported.

kshepherd commented on dspace/config/modules/requestitem.cfg:
```diff
@@ -0,0 +1,60 @@
+# Valid access periods, in seconds, to allow the approver to select when using links.
+# These are presented to users with friendly labels via i18n keys in the angular frontend.
+# These should be in a format like +<n><unit>, where <n> is the number of units and <unit> is one of
+# "SECONDS", "MINUTES", "HOURS", "DAYS", "WEEKS", "MONTHS", "YEARS". (or singular if n=1)
```
thanks, done

tdonohue reviewed (approved):
👍 Thanks @kshepherd !  This now looks correct to me, and the `1 week` option works properly (just re-tested it).

So, I'm now +1 on all these PRs.  They are just waiting on a second +1 from either a tester or another code reviewer.  Maybe @ybnd or @paulo-graca can help us out? (as both of them previously reviewed & I believe their feedback has been addressed)

ybnd commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/RequestItemRepository.java:
```diff
@@ -282,33 +328,45 @@ public RequestItemRest put(Context context, HttpServletRequest request,
+        if (!requestItem.isAccept_request() ||
+                requestItem.getAccess_expiry() == null ||
+                requestItem.getAccess_expiry().isBefore(Instant.now())) {
+            throw new AccessDeniedException("Access has not been granted for this item request");
```
Would it be possible to still expose expired `RequestItem`s, maybe with a `boolean expired` flag based on the `isBefore` above?

This would make things a lot more clear for people that try to visit an expired link; right now they will just see a login page.

The Bitstreams would still be "safe" ~ `authorizeAccessByAccessToken`

(related to https://github.com/DSpace/dspace-angular/pull/3984#discussion_r2016742001 and https://github.com/DSpace/dspace-angular/pull/3984#discussion_r2016745272)

kshepherd replied:
I've done this, and changed the frontnend to display different error alerts depending on the results. Thanks!

ybnd reviewed (changes requested):
One question/suggestion regarding expired access tokens

kshepherd committed:
Request-a-copy: Better error / auth flag handling

* New "isAccessExpired" flag for REST model
  (set by converter before responding)
* No changes to bitstream handling - that
  is still a hard auth check which fails
  on any authZ failure or error

ybnd reviewed (approved):
Perfect, thank you!

tdonohue commented:
Merging as this is now at +2

saschaszott commented:
@kshepherd , the new finder method `byAccessToken` which was added by this PR is not included in REST contract 👉 https://github.com/DSpace/RestContract/issues/348
